cds_ils/migrator/series/api.py

```diff
@@ -66,75 +72,109 @@ def create_multipart_volumes(pid, multipart_legacy_recid, migration_volumes):
-                            key, multipart_legacy_recid
+        if "isbn" in obj:
+            if "isbns" not in volume:
+                volume["isbns"] = []
+            volume["isbns"].append({
+                "value": obj["isbn"],
+                "is_electronic": bool(obj["is_electronic"])
+            })
+            # TODO physical description
+        elif "barcode" in obj:
+            # this entry represents an item
+            if "items" not in volume:
+                volume["items"] = []
+            volume["items"].append({
+                "barcode": obj["barcode"]
+            })
+        else:
+            for key in obj:
+                if key != "volume":
+                    if key in volume:
+                        raise KeyError(
+                            'Duplicate key "{}" for multipart {}'.format(
+                                key, multipart_legacy_recid
+                            )
                         )
-                    )
-                volume[key] = obj[key]
+                    volume[key] = obj[key]
 
     volume_numbers = iter(sorted(volumes.keys()))
 
-    # Re-use the current record for the first volume
-    # TODO review this - there are more cases of multiparts
-    first_volume = next(volume_numbers)
-    first = Document.get_record_by_pid(pid)
-    if "title" in volumes[first_volume]:
-        first["title"] = volumes[first_volume]["title"]
-        first["volume"] = first_volume
-    first["_migration"]["multipart_legacy_recid"] = multipart_legacy_recid
+    blueprint = document_blueprint
+    blueprint["_migration"]["multipart_legacy_recid"] = multipart_legacy_recid
+    blueprint["authors"] = blueprint["_migration"]["authors"] \
+        if "authors" in blueprint["_migration"] else []
+    blueprint["serial_title"] = blueprint.get("title")
+
     # to be tested
-    if "legacy_recid" in first:
-        del first["legacy_recid"]
-    first.commit()
-    yield first
+    if "legacy_recid" in blueprint:
+        del blueprint["legacy_recid"]
 
     # Create new records for the rest
     for number in volume_numbers:
-        temp = first.copy()
-        temp["title"] = volumes[number]["title"]
+        volume = volumes[number]
+        temp = blueprint.copy()
+        if "title" in volume and volume["title"]:
+            temp["title"] = volume["title"]
         temp["volume"] = number
         record_uuid = uuid.uuid4()
-        provider = DocumentIdProvider.create(
-            object_type="rec", object_uuid=record_uuid
-        )
-        temp["pid"] = provider.pid.pid_value
-        record = Document.create(temp, record_uuid)
-        record.commit()
-        yield record
+        try:
+            provider = DocumentIdProvider.create(
+                object_type="rec", object_uuid=record_uuid
+            )
+            temp["pid"] = provider.pid.pid_value
+            record = Document.create(temp, record_uuid)
+            record.commit()
```

Comment:
wrap these lines in `with db.session.begin_nested():` otherwise we have created PID but no document, PID will be empty, with no metadata

also where do we perform db.session.commit() for these new records?

Reply:
It was missing indeed, I suppose the autocommit took over.